File: Data Structures & Algorithms/binary-tree-diameter/submission-0.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    result: number = 0
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root: TreeNode | null) {
        /**
         for a given node, the # of edges in the node is
         the height of the left subtree + the height of the right subtree

         using the 1st example on the left
         node 5 has height of 0
         node 3 has height of 1
         node 4 has height of 0
         
         node 2 height = 
         max (left subtree, right subtree)
         node 2 diameter = left (2) + right (1)
         */
        this.findHeight(root)
        return this.result
    }

    findHeight(node: TreeNode | null): number {
        if (node === null) return 0

        let left = this.findHeight(node.left)
        let right = this.findHeight(node.right) 

        this.result = Math.max(this.result, left + right)

        return 1 + Math.max(left, right)
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        """
            at each node 
                get the height of the left subtree
                get the height of the right subtree
                caluclate the difference
                    if > 1
                        return false
        """
        if not root:
            return True

        def findH(node):
            if not node:
                return (0, True)
        
            left = findH(node.left)
            right = findH(node.right)

            if not left[1] or not right[1]:
                return (0, False)

            difference = abs(left[0] - right[0])
            if difference > 1:
                return (0, False)
            return (1 + max(left[0], right[0]), True)
        
        result = findH(root)
        return result[1]